Public/Src/Engine/Processes/FileAccessManifest.cs

```diff
@@ -504,17 +728,86 @@ public ArraySegment<byte> GetPayloadBytes(FileAccessSetup setup, MemoryStream st
                 WriteTranslationPathStrings(writer, DirectoryTranslator);
                 WriteErrorDumpLocation(writer, InternalDetoursErrorNotificationFile);
-                WriteFlagsBlock(writer);
+                WriteFlagsBlock(writer, m_fileAccessManifestFlag);
                 WriteExtraFlagsBlock(writer, FileAccessManifestExtraFlag.None);
                 WritePipId(writer, PipId);
                 WriteReportBlock(writer, setup);
                 WriteDllBlock(writer, setup);
-                WriteManifestTreeBlock(writer);
+                if (!IsManifestTreeBlockSealed)
+                {
+                    WriteManifestTreeBlock(writer);
+                }
+                else
+                {
+                    Contract.Assert(m_sealedManifestTreeBlock != null);
+                    writer.Write(m_sealedManifestTreeBlock);
+                }
 
                 return new ArraySegment<byte>(stream.GetBuffer(), 0, (int)stream.Position);
             }
         }
 
+        /// <summary>
+        /// Serializes this manifest.
+        /// </summary>
+        public void Serialize(Stream stream)
+        {
+            Contract.Requires(stream != null);
+
+            using (var writer = new BinaryWriter(stream, Encoding.Unicode, true))
+            {
+                WriteTranslationPathStrings(writer, DirectoryTranslator);
+                WriteErrorDumpLocation(writer, InternalDetoursErrorNotificationFile);
+                WriteFlagsBlock(writer, m_fileAccessManifestFlag);
+                WritePipId(writer, PipId);
+                WriteChars(writer, m_messageCountSemaphoreName);
+
+                // The manifest tree block has to be serialized the last.
+                if (!IsManifestTreeBlockSealed)
+                {
+                    WriteManifestTreeBlock(writer);
+                }
+                else
+                {
+                    Contract.Assert(m_sealedManifestTreeBlock != null);
+                    writer.Write(m_sealedManifestTreeBlock);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Deserialize an instance of <see cref="FileAccessManifest"/> from stream.
+        /// </summary>
+        public static FileAccessManifest Deserialize(Stream stream)
+        {
+            Contract.Requires(stream != null);
+
+            using (var reader = new BinaryReader(stream, Encoding.Unicode, true))
+            {
+                DirectoryTranslator directoryTranslator = ReadTranslationPathStrings(reader);
+                string internalDetoursErrorNotificationFile = ReadErrorDumpLocation(reader);
+                FileAccessManifestFlag fileAccessManifestFlag = ReadFlagsBlock(reader);
+                long pipId = ReadPipId(reader);
+                string messageCountSemaphoreName = ReadChars(reader);
+
+                byte[] sealedManifestTreeBlock;
+                using (MemoryStream ms = new MemoryStream(4096))

```

Comment:
>4096 [](start = 58, length = 4)

Probably wise to file a workitem to see about perf. a) If this is a good buffer size, b) to see if the buffers should be pooled, now they are just allocagted and thrown away

Reply:
Mike doesn't seem to be fond of introducing work items. I'll add TODO instead.

---
In reply to: [276815177](https://github.com/Microsoft/BuildXL/pull/159#discussion_r276815177) [](ancestors = 276815177)